internal/uvm/virtual_device.go

```diff
@@ -10,57 +11,110 @@ import (
 	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
 )
 
-// VPCIDevice represents a vpci device. Holds its guid and a handle to the uvm it
-// belongs to.
+// VPCIDevice represents a vpci device.
 type VPCIDevice struct {
+	// vm is the handle to the UVM that this device belongs to
 	vm *UtilityVM
-	ID string
+	// VMBusGUID is the instance ID for this device when it is exposed via VMBus
+	VMBusGUID string
+	// deviceInstanceID is the instance ID of the device on the host
+	deviceInstanceID string
+	// refCount stores the number of references to this device in the UVM
+	refCount uint32
 }
 
 // Release frees the resources of the corresponding vpci device
 func (vpci *VPCIDevice) Release(ctx context.Context) error {
-	if err := vpci.vm.RemoveDevice(ctx, vpci.ID); err != nil {
+	if err := vpci.vm.removeDevice(ctx, vpci.deviceInstanceID); err != nil {
 		return fmt.Errorf("failed to remove VPCI device: %s", err)
 	}
 	return nil
 }
 
-// AssignDevice assigns a new vpci device to the uvm
-func (uvm *UtilityVM) AssignDevice(ctx context.Context, device hcsschema.VirtualPciDevice) (*VPCIDevice, error) {
+// AssignDevice assigns a vpci device to the uvm
+// if the device already exists, the stored VPCIDevice's ref count is increased
+// and the VPCIDevice is returned.
+// Otherwise, a new request is made to assign the target device indicated by the deviceID
+// onto the UVM. A new VPCIDevice entry is made on the UVM and the VPCIDevice is returned
+// to the caller
+func (uvm *UtilityVM) AssignDevice(ctx context.Context, deviceID string) (*VPCIDevice, error) {
+	if uvm.operatingSystem == "windows" {
+		return nil, errors.New("assigned devices is not currently supported on wcow")
+	}
+
 	guid, err := guid.NewV4()
 	if err != nil {
 		return nil, err
 	}
-	id := guid.String()
+	vmBusGUID := guid.String()
 
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	if err := uvm.modify(ctx, &hcsschema.ModifySettingRequest{
-		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, id),
+
+	existingVPCIDevice := uvm.vpciDevices[deviceID]
+	if existingVPCIDevice != nil {
+		existingVPCIDevice.refCount++
+		return existingVPCIDevice, nil
+	}
+
+	targetDevice := hcsschema.VirtualPciDevice{
+		Functions: []hcsschema.VirtualPciFunction{
+			{
+				DeviceInstancePath: deviceID,
+			},
+		},
+	}
+
+	request := &hcsschema.ModifySettingRequest{
+		ResourcePath: fmt.Sprintf(virtualPciResourceFormat, vmBusGUID),
 		RequestType:  requesttype.Add,
-		Settings:     device,
-		GuestRequest: guestrequest.GuestRequest{
+		Settings:     targetDevice}
+
+	if uvm.operatingSystem != "windows" {
```

Comment:
I think it would still be good to add a comment here saying "WCOW (when supported) does not require a guest request as part of assigning the device".

Reply:
I think that would be best left as a comment for the hyper-v isolated assigned devices support PR. Thoughts?

Reply:
Does that PR exist currently? I just want to avoid any confusion over seeing this code and it not being clear why we don't do the guest request for WCOW. If something is coming imminently that will resolve that issue, that sounds fine to me.

Reply:
It does not. I guess it's reasonable to add a comment here since I do have this extra check.